File: src/info/dbinfo.rs
use sqlx::{ Connect, Error, Cursor, row::Row };
use sqlx::sqlite::SqliteConnection;
use std::collections::HashMap;
use crate::info::tableinfo::TableInfo;

pub struct DbInfo {
    name: String,
    tables: HashMap<String, TableInfo>,
}


impl DbInfo {
    pub async fn load(&mut self, path: &str) -> Result<(), Error> {
        self.name = path.to_string();
        let mut connection = SqliteConnection::connect(self.get_url()).await?;
        let mut rows = sqlx::query("SELECT name FROM sqlite_master WHERE type ='table' AND name NOT LIKE 'sqlite_%';")
                            .fetch(&mut connection);
        let mut table_names: Vec<String> = Vec::new();  // Load the table names first so we can re-use the connection
        while let Some(row) = rows.next().await? {

            let name: &str = row.get(0);
            table_names.push(name.to_string());
        } 
        // Now load each table info                       
        for table_name in table_names {
            let mut table: TableInfo = Default::default();
            table.load_from_db(&table_name, &mut connection).await?;
            self.tables.insert(table_name, table);    
        }
        Ok(())
    }

    fn get_url(&self) -> String {
        String::from("sqlite:") + &self.name
    }

    async fn compare_tables(&self, table_name: &str, num_columns: usize, db2: &DbInfo) -> Result<(bool, usize), Error> {
        let mut matches = true;
        let mut connection = SqliteConnection::connect(self.get_url()).await?;
        let mut connection2 = SqliteConnection::connect(db2.get_url()).await?;
        let sql = format!("SELECT * FROM {}", &table_name);
        let mut rows = sqlx::query(&sql).fetch(&mut connection);
        let mut rows2 = sqlx::query(&sql).fetch(&mut connection2);  
        let mut row_num: usize = 0;  
        while let Some(row) = rows.next().await? {
            if let Some(row2) = rows2.next().await? {
                for col in 0..num_columns {
                    let cell: Option<&str> = row.get(col);
                    let cell2: Option<&str> = row2.get(col);
                    if  cell != cell2 {
                        matches = false;
                        break;
                    }
                }
                if matches == false {
                    break;
                } else {
                    row_num += 1;
                }
            } else {
                matches = false;
                break;
            }
        }

        Ok((matches, row_num))
    }

    pub async fn compare(&self, db2: &DbInfo) -> Result<bool, Error> {
        let mut matches = true;

        let mut matched: Vec<&TableInfo> = Vec::new();
        let mut only_db1: Vec<&str> = Vec::new();
        let mut only_db2: Vec<&str> = Vec::new();
    
        for (table_name, table_info) in &self.tables {
            if let Some(table2) = db2.tables.get(table_name) {
                if !table_info.compare_tables(table2) {
                    println!("Different Table: {} {} {} but in {} {}", &table_name, &self.name, table_info.display_str(false), 
                            &db2.name, table2.display_str(false));
                    matches = false;
                } else {
                    if table_info.get_num_rows() < 1000 {
                        let (res, row_num) = self.compare_tables(&table_name, table_info.get_num_columns(), db2).await?;
                        if !res {
                            println!("Different Data for Table: {} in row {}", &table_name, row_num + 1);
                            matches = false;
                        } else {
                            matched.push(&table_info);
                        }
                    } else {
                        matched.push(&table_info);
                    }
                }
            } else {
                only_db1.push(&table_name);
                matches = false;
            }
        }
        for (table_name, _table_info) in &db2.tables {
            if !self.tables.contains_key(table_name) {
                only_db2.push(&table_name);
                matches = false;
            }
        }

        if matched.len() > 0 {
            println!("*************** {} matched tables ****************", matched.len());
            for table in matched {
                println!("Table: {} ", table.display_str(true));
            }
        }

        if only_db1.len() > 0 {
            println!("*************** {} tables only in {} ****************", only_db1.len(), &self.name);
            for table in only_db1 {
                println!("Table: {} ", table);
            }
        }

        if only_db2.len() > 0 {
            println!("*************** {} tables only in {} ****************", only_db2.len(), &db2.name);
            for table in only_db2 {
                println!("Table: {} ", table);
            }
        }

        Ok(matches)
    }
}

impl Default for DbInfo {
    fn default() -> Self {
        DbInfo {
            name: String::from(""),
            tables: HashMap::new(),
        }
    }
}

File: Java/src/main/java/dbComp/DbInfo.java
package dbComp;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.AbstractMap;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.Vector;
import java.util.Map.Entry;

class DbInfo
{
    String name;
    HashMap<String, TableInfo> tables = new HashMap<String, TableInfo>();
    Connection sqlconn;
    public DbInfo(String path) throws SQLException
    {
        this.name = path;
        this.sqlconn = DriverManager.getConnection("jdbc:sqlite:" + path);

        Statement stmt  = sqlconn.createStatement();
        String sql = "SELECT name FROM sqlite_master WHERE type =\'table\' AND name NOT LIKE \'sqlite_%\'";
        ResultSet rs = stmt.executeQuery(sql);

        List<String> tableNames = new Vector<String>();
        while (rs.next())
        {
            tableNames.add(rs.getString(1));
        }
        for (String tableName : tableNames)
        {
            tables.put(tableName, new TableInfo(sqlconn, tableName));
        }
    }
    AbstractMap.SimpleEntry<Boolean, Integer> compareTables(String tableName, int numColumns, DbInfo db)
            throws SQLException
    {
        String sql = "SELECT * FROM '" + tableName + "'";
        Statement stmt  = sqlconn.createStatement();
        Statement stmt2  = db.sqlconn.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        ResultSet rs2 = stmt2.executeQuery(sql);

        int nRow = 0;
        while (rs.next())
        {
            rs2.next();
            for (int nCol = 0; nCol < numColumns; ++nCol)
            {
                if (!Objects.equals(rs.getObject(nCol + 1), rs2.getObject(nCol + 1)))
                    return new AbstractMap.SimpleEntry<Boolean, Integer>(false, nRow);
            }
            ++nRow;
        }
        return new AbstractMap.SimpleEntry<Boolean, Integer>(true, 0);
    }
    public boolean compare(DbInfo db) throws SQLException
    {
        List<TableInfo> matched = new Vector<TableInfo>();
        List<String> onlyDb1 = new Vector<String>();
        List<String> onlyDb2 = new Vector<String>();
        boolean matches = true;

        for (Entry<String, TableInfo> entry : tables.entrySet()) {
            TableInfo tableInfo = db.tables.get(entry.getKey());
            if (tableInfo != null) {
                if (!entry.getValue().compare(tableInfo)) {
                    System.out.printf("Different Table: %s %s %s but in %s %s\n", entry.getKey(), name, entry.getValue().getDescription(), 
                            db.name, tableInfo.getDescription());
                    matches = false;
                } else {
                    if (tableInfo.numRows < 1000) {
                        AbstractMap.SimpleEntry<Boolean, Integer> res = compareTables(entry.getKey(), tableInfo.numColumns(), db);
                        if (!res.getKey()) {
                            System.out.printf("Different Data for Table: %s in row %d\n", entry.getKey(), res.getValue() + 1);
                            matches = false;
                        } else {
                            matched.add(entry.getValue());
                        }
                    } else {
                        matched.add(entry.getValue());
                    }
                }
            } else {
                onlyDb1.add(entry.getKey());
                matches = false;
            }
        }
        for (Entry<String, TableInfo> entry : db.tables.entrySet()) {
            if (!tables.containsKey(entry.getKey())) {
                onlyDb2.add(entry.getKey());
                matches = false;
            }
        }

        if (matched.size() > 0) {
            System.out.printf("*************** %d matched tables ****************\n", matched.size());
            for (TableInfo table : matched) {
                System.out.printf("Table: %s\n", table.getDescription());
            }
        }

        if (onlyDb1.size() > 0) {
            System.out.printf("*************** %d tables only in %s ****************\n", onlyDb1.size(), name);
            for (String table : onlyDb1) {
                System.out.printf("Table: %s\n", table);
            }
        }

        if (onlyDb2.size() > 0) {
            System.out.printf("*************** %d tables only in %s ****************\n", onlyDb2.size(), db.name);
            for (String table : onlyDb2) {
                System.out.printf("Table: %s\n", table);
            }
        }

        return matches;
    }
}
